cmd/snap/inhibit.go

```diff
@@ -34,143 +34,153 @@ import (
-	if err != nil {
-		return err
-	}
-	if hint == runinhibit.HintNotInhibited {
-		return nil
+	flow := newInhibitionFlow(snapName)
+	notified := false
+	inhibited := func(hint runinhibit.Hint, inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
+		if !notified {
+			// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+			// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+			// and then either unlocks it or changes to HintInhibitedForRefresh (see
+			// gateAutoRefreshHookHandler in hooks.go).
+			if hint != runinhibit.HintInhibitedForRefresh {
+				return false, nil
+			}
+			// Start notification flow.
+			if err := flow.Notify(); err != nil {
+				return true, err
+			}
+			// Make sure we call notification flow only once.
+			notified = true
+		}
+		return false, nil
 	}
 
-	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
-	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
-	// and then either unlocks it or changes to HintInhibitedForRefresh (see
-	// gateAutoRefreshHookHandler in hooks.go).
-	// waitInhibitUnlock will return also on HintNotInhibited.
-	notInhibited, err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh)
+	hintFlock, err := runinhibitWaitWhileInhibited(snapName, nil, inhibited, 500*time.Millisecond)
 	if err != nil {
+		// It is fine to return an error here without finishing the notification
+		// flow because we either failed because of it or before it, so it
+		// should not have started in the first place.
 		return err
 	}
-	if notInhibited {
-		return nil
-	}
 
-	if isGraphicalSession() {
-		notifiedDesktopIntegration, err := tryNotifyRefreshViaSnapDesktopIntegrationFlow(snapName)
-		if err != nil {
+	// XXX: closing as we don't need it for now, this lock will be used in a later iteration
+	hintFlock.Close()
+
+	if notified {
+		if err := flow.Finish(); err != nil {
 			return err
 		}
-		if notifiedDesktopIntegration {
-			return nil
-		}
-		return graphicalSessionFlow(snapName, hint)
 	}
-	// terminal and headless
-	return textFlow(snapName, hint)
+
+	return nil
+}
+
+type inhibitionFlow interface {
+	Notify() error
+	Finish() error
 }
 
-func inhibitMessage(snapName string, hint runinhibit.Hint) string {
-	switch hint {
-	case runinhibit.HintInhibitedForRefresh:
-		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
-	default:
-		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+var newInhibitionFlow = func(snapName string) inhibitionFlow {
+	if isGraphicalSession() {
+		return &graphicalFlow{snapName: snapName}
 	}
+	return &textFlow{snapName: snapName}
 }
 
-var isGraphicalSession = func() bool {
-	// TODO: uncomment once there is a proper UX review
-	//return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
-	return false
+type textFlow struct {
+	snapName string
+
+	pb progress.Meter
```

Comment:
Is this accessed from different goroutines?

Reply:
No, this is only accessed through `snap-run`